File: backend/app/db/repositories/matches.py
# -*- coding: utf-8 -*-
"""
CRUD operations for the Matches table.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional, Union

from loguru import logger

from app.db.session import transaction, with_retry


@with_retry(max_retries=3, retry_delay=0.5)
def create_match_by_team_ids(
    winner_team_id: int,
    loser_team_id: int,
    played_at: Union[datetime, str],
    is_fanny: bool = False,
    notes: Optional[str] = None,
) -> Optional[int]:
    """
    Record a new match in the database.

    Parameters
    ----------
    winner_team_id : int
        ID of the winning team
    loser_team_id : int
        ID of the losing team
    played_at : Union[datetime, str]
        Date and time when the match was played
    is_fanny : bool, optional
        Whether the match was a fanny (10-0), by default False
    notes : Optional[str], optional
        Optional notes about the match, by default None

    Returns
    -------
    Optional[int]
        ID of the newly created match, or None on failure
    """
    try:
        # ##: Convert string to datetime if necessary.
        if isinstance(played_at, str):
            played_at = datetime.fromisoformat(played_at)

        # ##: Validate teams are different.
        if winner_team_id == loser_team_id:
            logger.error("Winner and loser team IDs cannot be the same")
            return None

        with transaction() as db_client:
            match_data = {
                "winner_team_id": winner_team_id,
                "loser_team_id": loser_team_id,
                "is_fanny": is_fanny,
                "played_at": played_at.isoformat() if isinstance(played_at, datetime) else played_at,
            }
            if notes is not None:
                match_data["notes"] = notes

            response = db_client.table("matches").insert(match_data).execute()

            if response.data and len(response.data) > 0:
                new_match_id = response.data[0].get("match_id")
                if new_match_id is not None:
                    logger.info(f"Match created successfully with ID: {new_match_id}")
                    return new_match_id

            logger.warning("Match creation did not return an ID or failed.")
            return None
    except Exception as exc:
        logger.error(f"Failed to create match: {exc}")
        return None


@with_retry(max_retries=3, retry_delay=0.5)
def get_match_by_id(match_id: int) -> Optional[Dict[str, Any]]:
    """
    Get a match by ID.

    Parameters
    ----------
    match_id : int
        ID of the match to retrieve

    Returns
    -------
    Optional[Dict[str, Any]]
        Match data as a dictionary, or None if not found
    """
    with transaction() as db_client:
        response = (
            db_client.table("matches")
            .select("match_id, winner_team_id, loser_team_id, is_fanny, played_at, notes")
            .eq("match_id", match_id)
            .maybe_single()
            .execute()
        )
        if response and response.data:
            if response.data.get("played_at"):
                response.data["played_at"] = datetime.fromisoformat(response.data["played_at"])
            return response.data
    return None


@with_retry(max_retries=3, retry_delay=0.5)
def get_matches_by_team_id(
    team_id: int,
    limit: int = 100,
    offset: int = 0,
    is_fanny: bool = False,
    start_date: Optional[Union[datetime, str]] = None,
    end_date: Optional[Union[datetime, str]] = None,
) -> List[Dict[str, Any]]:
    """
    Get all matches involving a specific team.

    Parameters
    ----------
    team_id : int
        ID of the team
    limit : int, optional
        Maximum number of matches to return (default: 100).
    offset : int, optional
        Number of matches to skip for pagination (default: 0).
    is_fanny : bool, optional
        Whether to filter matches by fanny (10-0), by default False
    start_date : Optional[Union[datetime, str]], optional
        Start date to filter matches (inclusive), by default None
    end_date : Optional[Union[datetime, str]], optional
        End date to filter matches (inclusive), by default None

    Returns
    -------
    List[Dict[str, Any]]
        List of match dictionaries
    """
    try:
        # ##: Convert string dates to datetime objects.
        if isinstance(start_date, str):
            start_date = datetime.fromisoformat(start_date)
        if isinstance(end_date, str):
            end_date = datetime.fromisoformat(end_date)
        if end_date is not None:
            end_date = datetime.combine(end_date, datetime.max.time())

        with transaction() as db_client:
            response = db_client.rpc(
                "get_team_match_history",
                {
                    "p_team_id": team_id,
                    "p_limit": limit,
                    "p_offset": offset,
                    "p_is_fanny": is_fanny,
                    "p_start_date": start_date,
                    "p_end_date": end_date,
                },
            ).execute()

        matches_data = []
        if response.data:
            for row_item in response.data:
                row_item["played_at"] = (
                    datetime.fromisoformat(row_item["played_at"]) if row_item["played_at"] else None
                )
                matches_data.append(row_item)
        return matches_data
    except Exception as exc:
        logger.error(f"Failed to get matches for team ID {team_id}: {exc}")
        return []


@with_retry(max_retries=3, retry_delay=0.5)
def get_matches_by_player_id(
    player_id: int,
    limit: int = 100,
    offset: int = 0,
    is_fanny: bool = False,
    start_date: Optional[Union[datetime, str]] = None,
    end_date: Optional[Union[datetime, str]] = None,
) -> List[Dict[str, Any]]:
    """
    Get all matches involving a specific player with optional filtering and pagination.

    Parameters
    ----------
    player_id : int
        ID of the player
    limit : int, optional
        Maximum number of matches to return, by default 100
    offset : int, optional
        Number of matches to skip, by default 0
    is_fanny : bool, optional
        Whether to filter matches by fanny (10-0), by default False
    start_date : Optional[Union[datetime, str]], optional
        Start date to filter matches (inclusive), by default None
    end_date : Optional[Union[datetime, str]], optional
        End date to filter matches (inclusive), by default None

    Returns
    -------
    List[Dict[str, Any]]
        List of match dictionaries including match details and player's ELO change
    """
    try:
        # ##: Convert string dates to datetime objects.
        if isinstance(start_date, str):
            start_date = datetime.fromisoformat(start_date)
        if isinstance(end_date, str):
            end_date = datetime.fromisoformat(end_date)
        if end_date is not None:
            end_date = datetime.combine(end_date, datetime.max.time())

        with transaction() as db_client:
            response = db_client.rpc(
                "get_player_matches_json",
                {
                    "p_end_date": end_date,
                    "p_is_fanny": is_fanny,
                    "p_limit": limit,
                    "p_offset": offset,
                    "p_player_id": player_id,
                    "p_start_date": start_date,
                },
            ).execute()

        matches_details = []
        if response.data:
            for row_item in response.data:
                row_item["played_at"] = (
                    datetime.fromisoformat(row_item["played_at"]) if row_item["played_at"] else None
                )
                matches_details.append(row_item)
            return matches_details
    except Exception as exc:
        logger.error(f"Failed to get matches for player {player_id}: {exc}")
        return []


@with_retry(max_retries=3, retry_delay=0.5)
def get_all_matches(limit: int = 100, offset: int = 0) -> List[Dict[str, Any]]:
    """
    Get all matches with pagination support.

    Parameters
    ----------
    limit : int, optional
        Maximum number of matches to return, by default 100
    offset : int, optional
        Number of matches to skip, by default 0

    Returns
    -------
    List[Dict[str, Any]]
        List of match dictionaries
    """
    try:
        with transaction() as db_client:
            response = db_client.rpc("get_all_matches_with_details", {"p_limit": limit, "p_offset": offset}).execute()
            matches_list = []

        if response.data:
            for row_data in response.data:
                row_data["played_at"] = datetime.fromisoformat(row_data["played_at"])
                matches_list.append(row_data)
        return matches_list
    except Exception as exc:
        logger.error(f"Failed to get all matches: {exc}")
        return []


@with_retry(max_retries=3, retry_delay=0.5)
def delete_match_by_id(match_id: int) -> bool:
    """
    Delete a match from the database.

    Parameters
    ----------
    match_id : int
        ID of the match to delete

    Returns
    -------
    bool
        True if the deletion was successful, False otherwise
    """
    try:
        with transaction() as db_client:
            response = db_client.table("matches").delete(count="exact").eq("match_id", match_id).execute()
            if response.count is not None and response.count > 0:
                logger.info(f"Match with ID {match_id} deleted successfully.")
                return True
            logger.warning(f"Match with ID {match_id} not found or not deleted.")
            return False
    except Exception as exc:
        logger.error(f"Failed to delete match ID {match_id}: {exc}")
        return False

File: lib/db/repositories/matches.ts
// [>]: Match repository - all database operations for matches table.
// Uses RPC functions for complex queries with team/player data.

import { getSupabaseClient } from "@/lib/db/client";
import { withRetry } from "@/lib/db/retry";
import {
  MatchNotFoundError,
  MatchCreationError,
  MatchOperationError,
  MatchDeletionError,
} from "@/lib/errors/api-errors";

// [>]: Database row type for raw Supabase responses.
interface MatchDbRow {
  match_id: number;
  winner_team_id: number;
  loser_team_id: number;
  is_fanny: boolean;
  played_at: string;
  notes?: string | null;
}

// [>]: Extended row type with team details from RPC.
interface MatchWithTeamsRow extends MatchDbRow {
  winner_team?: {
    team_id: number;
    player1_id: number;
    player2_id: number;
    global_elo: number;
  } | null;
  loser_team?: {
    team_id: number;
    player1_id: number;
    player2_id: number;
    global_elo: number;
  } | null;
  // [>]: ELO changes keyed by player ID (from get_player_matches_json RPC).
  elo_changes?: Record<
    string,
    { old_elo: number; new_elo: number; difference: number }
  > | null;
}

// [>]: Query options for match filtering.
interface MatchQueryOptions {
  limit?: number;
  offset?: number;
  isFanny?: boolean;
  startDate?: string;
  endDate?: string;
}

// [>]: Create a new match record. Returns the match_id.
// Does NOT update ELOs - that is the service layer's responsibility.
async function createMatchByTeamIdsImpl(
  winnerTeamId: number,
  loserTeamId: number,
  playedAt: string,
  isFanny: boolean = false,
  notes?: string,
): Promise<number> {
  const client = getSupabaseClient();

  const matchData: Record<string, unknown> = {
    winner_team_id: winnerTeamId,
    loser_team_id: loserTeamId,
    played_at: playedAt,
    is_fanny: isFanny,
  };

  if (notes) {
    matchData.notes = notes;
  }

  const { data, error } = await client
    .from("matches")
    .insert(matchData)
    .select("match_id")
    .single();

  if (error) {
    throw new MatchCreationError(`Failed to create match: ${error.message}`);
  }

  if (!data?.match_id) {
    throw new MatchCreationError("Failed to create match: no ID returned");
  }

  return data.match_id;
}

// [>]: Lookup match by ID. Throws MatchNotFoundError if not found.
async function getMatchByIdImpl(matchId: number): Promise<MatchDbRow> {
  const client = getSupabaseClient();

  const { data, error } = await client
    .from("matches")
    .select(
      "match_id, winner_team_id, loser_team_id, is_fanny, played_at, notes",
    )
    .eq("match_id", matchId)
    .maybeSingle();

  if (error) {
    throw new MatchOperationError(`Database error: ${error.message}`);
  }

  if (!data) {
    throw new MatchNotFoundError(matchId);
  }

  return data;
}

// [>]: Get all matches with pagination and optional filters (uses RPC for joined data).
async function getAllMatchesImpl(
  options: MatchQueryOptions = {},
): Promise<MatchWithTeamsRow[]> {
  const { limit = 100, offset = 0, startDate, endDate, isFanny } = options;
  const client = getSupabaseClient();

  const { data, error } = await client.rpc("get_all_matches_with_details", {
    p_limit: limit,
    p_offset: offset,
    p_start_date: startDate ?? null,
    p_end_date: endDate ?? null,
    p_is_fanny: isFanny ?? null,
  });

  if (error) {
    throw new MatchOperationError(
      `Failed to get all matches: ${error.message}`,
    );
  }

  return data ?? [];
}

// [>]: Get matches filtered by team ID (uses RPC).
async function getMatchesByTeamIdImpl(
  teamId: number,
  options: MatchQueryOptions = {},
): Promise<MatchWithTeamsRow[]> {
  const { limit = 100, offset = 0, isFanny, startDate, endDate } = options;
  const client = getSupabaseClient();

  const { data, error } = await client.rpc("get_team_match_history", {
    p_team_id: teamId,
    p_limit: limit,
    p_offset: offset,
    p_is_fanny: isFanny ?? null,
    p_start_date: startDate ?? null,
    p_end_date: endDate ?? null,
  });

  if (error) {
    throw new MatchOperationError(
      `Failed to get matches for team: ${error.message}`,
    );
  }

  return data ?? [];
}

// [>]: Get matches filtered by player ID (uses RPC).
async function getMatchesByPlayerIdImpl(
  playerId: number,
  options: MatchQueryOptions = {},
): Promise<MatchWithTeamsRow[]> {
  const { limit = 100, offset = 0, isFanny, startDate, endDate } = options;
  const client = getSupabaseClient();

  const { data, error } = await client.rpc("get_player_matches_json", {
    p_player_id: playerId,
    p_limit: limit,
    p_offset: offset,
    p_is_fanny: isFanny ?? null,
    p_start_date: startDate ?? null,
    p_end_date: endDate ?? null,
  });

  if (error) {
    throw new MatchOperationError(
      `Failed to get matches for player: ${error.message}`,
    );
  }

  return data ?? [];
}

// [>]: Delete match by ID. Throws MatchNotFoundError if not found.
async function deleteMatchByIdImpl(matchId: number): Promise<void> {
  const client = getSupabaseClient();

  const { count, error } = await client
    .from("matches")
    .delete({ count: "exact" })
    .eq("match_id", matchId);

  if (error) {
    throw new MatchDeletionError(`Failed to delete match: ${error.message}`);
  }

  if (count === 0) {
    throw new MatchNotFoundError(matchId);
  }
}

// [>]: Export wrapped functions with retry logic.
export const createMatchByTeamIds = withRetry(createMatchByTeamIdsImpl);
export const getMatchById = withRetry(getMatchByIdImpl);
export const getAllMatches = withRetry(getAllMatchesImpl);
export const getMatchesByTeamId = withRetry(getMatchesByTeamIdImpl);
export const getMatchesByPlayerId = withRetry(getMatchesByPlayerIdImpl);
export const deleteMatchById = withRetry(deleteMatchByIdImpl);

// [>]: Export types for use in services.
export type { MatchDbRow, MatchWithTeamsRow, MatchQueryOptions };
